.github/workflows/ci.yml

```diff
@@ -0,0 +1,119 @@
+            scan-docker-image: 'security-actions/scan-docker-image/**'
+            sca: 'security-actions/sca/**'
+            scan-rust: 'security-actions/scan-rust/**'
+            semgrep: 'security-actions/semgrep/**'
+            sign-docker-image: 'security-actions/sign-docker-image/**'
+            code-build-actions: 'code-build-actions/build-js-sdk/**'
+            lua-lint: 'code-check-actions/lua-lint/**'
+            rust-lint: 'code-check-actions/rust-lint/**'
+            slack-actions: 'slack-actions/workflow-notification/**'
+            cleanup: 'pr-previews/cleanup/**'
+            up-to-date: 'pr-previews/up-to-date/**'
+            validate: 'pr-previews/validate/**'
+
+  setup-dependencies:
+    needs: changes
+    runs-on: ubuntu-latest
+    if: ${{ needs.changes.outputs.packages != '[]' }}
+    outputs:
+      nodeVersion: ${{ steps.node-version.outputs.nodeVersion }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Setup Node
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
+
+      - name: Node Version
+        id: node-version
+        run: |
+          nodeVersion="$(node --version | grep -oE '[0-9]+' | head -n1)"
+          echo "nodeVersion=$nodeVersion" >> $GITHUB_OUTPUT
+
+      - name: Install PNPM
+        run: |
+          npm install -g pnpm
+          pnpm --version
+          if [ -n "${{ github.head_ref }}" ]; then
+            git checkout ${{ github.head_ref }}
+          fi
+
+      - name: Dependency Cache
+        id: dependency-cache
+        uses: actions/cache@v3
+        with:
+          path: '**/node_modules'
+          key: pnpm-${{ steps.node-version.outputs.nodeVersion }}-${{ runner.os }}-${{ hashFiles('**/pnpm-lock.yaml') }}
+
+      - name: Install Dependencies
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' }}
+        run: |
+          pnpm install --frozen-lockfile
+
+      - name: Commit lint
+        uses: wagoid/commitlint-github-action@v5
+        with:
+          failOnWarnings: true
+
+  build:
+    needs: [ changes, setup-dependencies ]
+    if: ${{ needs.changes.outputs.packages != '[]' }}
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        package: ${{ fromJSON(needs.changes.outputs.packages) }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Setup Node
+        uses: actions/setup-node@v4
+        with:
+          node-version: 20
+
+      - name: Install PNPM
+        run: |
+          npm install -g pnpm
+          pnpm --version
+
+      - name: Dependency Cache
+        uses: actions/cache@v3
+        with:
+          path: '**/node_modules'
+          key: pnpm-${{ needs.setup-dependencies.outputs.nodeVersion }}-${{ runner.os }}-${{ hashFiles('**/pnpm-lock.yaml') }}
+
+      - name: Restore Dependencies
+        if: ${{ steps.dependency-cache.outputs.cache-hit != 'true' }}
```

Comment:
always reconstruct node_modules using `pnpm install --frozen-lockfile`. Pnpm and setup node will automatically optimize this.